File: ios/ZillitLCW/Core/Auth/EncryptionUtil.swift
import Foundation
import CommonCrypto

/// AES-256-CBC encryption for the moduledata header.
///
/// Matches backend `middleware/auth.js`:
/// - Algorithm: AES-256-CBC
/// - Manual PKCS5 padding (setAutoPadding(false) on backend)
/// - Output: **hex string** (not base64)
/// - Key: last 32 chars of ENCRYPTION_KEY
/// - IV: first 16 chars of IV_KEY
struct EncryptionUtil {

    // Derived from .env (same as backend)
    static let encryptionKey = "tuvwxyz123456789ABCDEFGHIJKLMNOP" // last 32 chars
    static let ivKey = "1234567890abcdef"                         // first 16 chars

    /// Encrypt plaintext using AES-256-CBC with manual PKCS5 padding.
    /// Returns a lowercase hex string (matching crypto-js / Node.js backend).
    static func encrypt(_ plainText: String) -> String? {
        guard let data = plainText.data(using: .utf8),
              let keyData = encryptionKey.data(using: .utf8),
              let ivData = ivKey.data(using: .utf8),
              keyData.count == 32,
              ivData.count == 16 else {
            return nil
        }

        // Apply PKCS5 padding manually
        let paddedData = applyPKCS5Padding(data)

        let bufferSize = paddedData.count + kCCBlockSizeAES128
        var buffer = Data(count: bufferSize)
        var numBytesEncrypted: size_t = 0

        let status = buffer.withUnsafeMutableBytes { bufferPtr in
            paddedData.withUnsafeBytes { dataPtr in
                keyData.withUnsafeBytes { keyPtr in
                    ivData.withUnsafeBytes { ivPtr in
                        CCCrypt(
                            CCOperation(kCCEncrypt),
                            CCAlgorithm(kCCAlgorithmAES),
                            0, // NO padding — we do it manually
                            keyPtr.baseAddress, kCCKeySizeAES256,
                            ivPtr.baseAddress,
                            dataPtr.baseAddress, paddedData.count,
                            bufferPtr.baseAddress, bufferSize,
                            &numBytesEncrypted
                        )
                    }
                }
            }
        }

        guard status == kCCSuccess else { return nil }
        buffer.count = numBytesEncrypted

        // Output as lowercase hex string (NOT base64)
        return buffer.map { String(format: "%02x", $0) }.joined()
    }

    /// Build the moduledata header payload and encrypt it.
    static func buildModuleDataHeader(userId: String, projectId: String, deviceId: String) -> String? {
        let payload: [String: Any] = [
            "device_id": deviceId,
            "project_id": projectId,
            "user_id": userId,
            "time_stamp": Int64(Date().timeIntervalSince1970 * 1000)
        ]

        guard let jsonData = try? JSONSerialization.data(withJSONObject: payload),
              let jsonString = String(data: jsonData, encoding: .utf8) else {
            return nil
        }

        return encrypt(jsonString)
    }

    // MARK: - PKCS5 Padding (matches backend)

    private static func applyPKCS5Padding(_ data: Data) -> Data {
        let blockSize = 16
        let padLength = blockSize - (data.count % blockSize)
        var padded = data
        padded.append(contentsOf: [UInt8](repeating: UInt8(padLength), count: padLength))
        return padded
    }
}

File: android/app/src/main/java/com/zillit/lcw/util/EncryptionUtil.kt
package com.zillit.lcw.util

import javax.crypto.Cipher
import javax.crypto.spec.IvParameterSpec
import javax.crypto.spec.SecretKeySpec

/**
 * EncryptionUtil — AES-256-CBC encryption for the moduledata header.
 *
 * Matches the backend's `middleware/auth.js`:
 * - Algorithm: AES-256-CBC
 * - Manual PKCS5 padding (autoPadding disabled on backend)
 * - Output: **HEX string** (not base64)
 * - Key: last 32 chars of ENCRYPTION_KEY
 * - IV: first 16 chars of IV_KEY
 */
object EncryptionUtil {

    /**
     * Encrypt plainText using AES-256-CBC with manual PKCS5 padding.
     * Returns a lowercase hex string (matching crypto-js / Node.js backend format).
     */
    fun encrypt(plainText: String, encryptionKey: String, ivKey: String): String? {
        return try {
            val keyBytes = encryptionKey.toByteArray(Charsets.UTF_8)
            val ivBytes = ivKey.toByteArray(Charsets.UTF_8)

            require(keyBytes.size == 32) { "Key must be 32 bytes" }
            require(ivBytes.size == 16) { "IV must be 16 bytes" }

            val keySpec = SecretKeySpec(keyBytes, "AES")
            val ivSpec = IvParameterSpec(ivBytes)

            // Apply PKCS5 padding manually (backend uses setAutoPadding(false))
            val plainBytes = plainText.toByteArray(Charsets.UTF_8)
            val paddedBytes = applyPKCS5Padding(plainBytes)

            // Encrypt with NoPadding (we padded manually)
            val cipher = Cipher.getInstance("AES/CBC/NoPadding")
            cipher.init(Cipher.ENCRYPT_MODE, keySpec, ivSpec)
            val encrypted = cipher.doFinal(paddedBytes)

            // Output as lowercase hex string (NOT base64)
            encrypted.joinToString("") { "%02x".format(it) }
        } catch (e: Exception) {
            e.printStackTrace()
            null
        }
    }

    /**
     * Decrypt a hex-encoded ciphertext using AES-256-CBC.
     */
    fun decrypt(hexCipherText: String, encryptionKey: String, ivKey: String): String? {
        return try {
            val keyBytes = encryptionKey.toByteArray(Charsets.UTF_8)
            val ivBytes = ivKey.toByteArray(Charsets.UTF_8)

            val keySpec = SecretKeySpec(keyBytes, "AES")
            val ivSpec = IvParameterSpec(ivBytes)

            // Decode hex to bytes
            val encryptedBytes = hexCipherText.chunked(2).map { it.toInt(16).toByte() }.toByteArray()

            val cipher = Cipher.getInstance("AES/CBC/NoPadding")
            cipher.init(Cipher.DECRYPT_MODE, keySpec, ivSpec)
            val decryptedPadded = cipher.doFinal(encryptedBytes)

            // Remove PKCS5 padding
            val decrypted = removePKCS5Padding(decryptedPadded)
            String(decrypted, Charsets.UTF_8)
        } catch (e: Exception) {
            e.printStackTrace()
            null
        }
    }

    /**
     * PKCS5 padding: pad to 16-byte block boundary.
     * Matches backend's `applyPKCS5Padding()`.
     */
    private fun applyPKCS5Padding(data: ByteArray): ByteArray {
        val blockSize = 16
        val padLength = blockSize - (data.size % blockSize)
        val padded = ByteArray(data.size + padLength)
        System.arraycopy(data, 0, padded, 0, data.size)
        for (i in data.size until padded.size) {
            padded[i] = padLength.toByte()
        }
        return padded
    }

    /**
     * Remove PKCS5 padding.
     * Matches backend's `removePKCS5Padding()`.
     */
    private fun removePKCS5Padding(data: ByteArray): ByteArray {
        if (data.isEmpty()) throw IllegalArgumentException("Empty buffer")
        val padByte = data.last().toInt() and 0xFF
        if (padByte < 1 || padByte > 16) throw IllegalArgumentException("Invalid padding byte: $padByte")
        return data.copyOfRange(0, data.size - padByte)
    }
}
